### ISSUE TITLE
FNFE: "File system element for parameter 'source' does not exist" when executing `DiktatSaveSmokeTest.beforeAll`

### ISSUE THREAD

**Issue by 0x6675636b796f75676974687562** [saveourtool/diktat#1476]
The problem started to occur after 54c8ccefc3009b8dd2a493c30590168318765b89, initially on my _TeamCity_ instance. Now, I encounter the  same problem trying to [perform a release](https://github.com/saveourtool/diktat/runs/7503805748):

```stacktrace
Error:  org.cqfn.diktat.ruleset.smoke.DiktatSaveSmokeTest  Time elapsed: 4.207 s  <<< ERROR!
java.io.FileNotFoundException: File system element for parameter 'source' does not exist: ''
	at org.apache.commons.io.FileUtils.requireExistsChecked(FileUtils.java:2719)
	at org.apache.commons.io.FileUtils.requireFileCopy(FileUtils.java:2751)
	at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:840)
	at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:756)
	at org.cqfn.diktat.ruleset.smoke.DiktatSaveSmokeTest$Companion.beforeAll$diktat_rules(DiktatSaveSmokeTest.kt:114)
	at org.cqfn.diktat.ruleset.smoke.DiktatSaveSmokeTest.beforeAll$diktat_rules(DiktatSaveSmokeTest.kt)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	...
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:55)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:223)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:175)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:139)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:456)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:169)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:595)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:581)
```

**Comment by 0x6675636b796f75676974687562** [comment]
We need the following enhancements done to the smoke tests:

1. introduce the retry mechanism (a fixed value of 5 attempts should be enough);
2. debug logging to stdout (e. g. "attempt %d of %d; downloading %s from %s to %s")
3. human readable failure messages. 

A few words on failure messages. Simple solutions like

```kotlin
val success = ...
Assert.assertTrue(success)
```

are only good until the 1st failure happens. After that, people unfamiliar with the internals of a specific test start wondering what exactly `true` and `false` are intended to mean:

![teamcity-test-failure](https://user-images.githubusercontent.com/73111822/180831347-d74bc99a-6e81-4eab-b58b-e8761382c9d9.png)

**Comment by petertrr** [comment]
@0x6675636b796f75676974687562 yes, we already have AssertJ as a dependency, so introducing better asertions is definitely a good idea.

However, the problem you are facing is more on the infrastructure side of things, because currently smoke test with SAVE is executed in the `test` phase, but relies on output of `package` of `diktat-ruleset` module. As a workaround (see both #1388 and #1463) we are simply splitting execution into `install -DskipTests` and `test` afterwards.

Actually I thought we don't run tests for release workflow, because if something is merged into master, then the tests should be passing.

**Comment by 0x6675636b796f75676974687562** [comment]
Fixed with #1477.
